natpmp/client.go

```diff
@@ -0,0 +1,204 @@
+	ExternalIP        net.IP
+}
+
+// ExternalAddress returns external IP address information from a NAT gateway,
+// as described in RFC 4886, section 3.2.
+func (c *Client) ExternalAddress(ctx context.Context) (*ExternalAddress, net.Addr, error) {
+	// This messages's request is always fixed, and the response always has a
+	// fixed size and response opcode. See:
+	// https://tools.ietf.org/html/rfc6886#section-3.2.
+	const (
+		size  = 12
+		reqOp = 0x00
+		resOp = 128
+	)
+
+	b := make([]byte, size)
+	n, addr, err := c.request(ctx, []byte{Version, reqOp}, b, resOp)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if n != size {
+		return nil, nil, io.ErrUnexpectedEOF
+	}
+
+	// We allocated a buffer internally, no need to produce a copy of the data
+	// for the output IP address.
+	return &ExternalAddress{
+		SinceStartOfEpoch: time.Duration(binary.BigEndian.Uint32(b[4:8])) * time.Second,
+		ExternalIP:        b[8:12],
+	}, addr, nil
+}
+
+// request serializes and implements backoff/retry for NAT-PMP request/response
+// interactions, as recommended by
+// https://tools.ietf.org/html/rfc6886#section-3.1.
+func (c *Client) request(ctx context.Context, req, res []byte, resOp int) (int, net.Addr, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer wg.Wait()
+
+	// Either wait for the parent context to be canceled or for this function to
+	// complete, and then unblock any outstanding reads and return control to
+	// the caller.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go func() {
+		defer wg.Done()
+		<-ctx.Done()
+		_ = c.pc.SetDeadline(time.Unix(0, 1))
+	}()
+
+	// Start with a 250ms delay on timeout error and double it up to 9 times
+	// per the RFC.
+	var nerr net.Error
+	timeout := 250 * time.Millisecond
+	for i := 0; i < 9; i++ {
+		if err := ctx.Err(); err != nil {
+			return 0, nil, err
+		}
+
+		// Send a request to the gateway and await its response or a timeout.
+		if err := c.pc.SetReadDeadline(time.Now().Add(timeout)); err != nil {
+			return 0, nil, err
+		}
+
+		if _, err := c.pc.WriteTo(req, c.gateway); err != nil {
+			return 0, nil, err
+		}
+
+		n, addr, err := c.pc.ReadFrom(res)
+		switch {
+		case errors.As(err, &nerr) && nerr.Timeout():
+			// Was this timeout produced by context cancelation? If so, return
+			// immediately. If not, double the timeout and retry.
+			if err := ctx.Err(); err != nil {
+				return 0, nil, err
+			}
+
+			timeout *= 2
+		case err == nil:
+			// Successful read, parse the header and verify the expected
+			// response opcode.
+			if err := checkHeader(res[:n], resOp); err != nil {
+				return n, addr, err
+			}
+
+			return n, addr, nil
+		default:
+			// Unexpected error.
+			return n, addr, err
+		}
+	}
+
+	// TODO(mdlayher): implement net.Error.Timeout?
+	return 0, nil, errors.New("natpmp: exhausted retries")
+}
+
+// checkHeader validates the header bytes of a NAT-PMP response.
+func checkHeader(b []byte, op int) error {
```

Comment:
Pursuant to what I said above about integrating with an external PacketConn, it would be very good to provide some external API for "IsPacketICareAbout([]byte) bool", so that an external demultiplexer can figure out if the packet is for this package or not. checkHeader would be a reasonable first implementation of said external API.

Reply:
SGTM.